### PR TITLE
Private link 4 monitor

### PR DESCRIPTION
**Pull request by VeryEarly**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json
Comment (arolshan) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: you forgot details which is an array of ErrorResponse.
can you please copy whatever's in the common

Comment (VeryEarly) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: this definition I put here is the same as ["ErrorResponse" in existing spec](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/monitor/resource-manager/Microsoft.Insights/stable/2015-04-01/autoscale_API.json#L793)

there will be conflict when I try to generate dotnet SDK if the definition is non-identical

Comment (arolshan) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: Okay, but we use the details property to expose more details on the error.

**Comment by VeryEarly** [comment]
Hi @arolshan ,

One question about the tag "2020-03" I added, since it's not the default tag, are there anyway to reference this specific tag while generating?

Or the question could be, can I change the default tag to "2020-03"?

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json
Comment (VeryEarly) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: I can add one more definition of something like "ErrorResponseExtension" to be inherited from "ErrorResponse" and have "details"/"additionalInfo", what do you think?

Comment (VeryEarly) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: correct me if I was wrong, the ErrorResponse would be referenced only in ErrorResponseException, does it matter if the ErrorResponse doesn't contain the "detial"/"additionalInfo"?

Comment (arolshan) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: This is how the error response object should look:

public class ErrorResponse
    {
        [JsonProperty("code")]
        public string Code { get; set; }

        [JsonProperty("message")]
        public string Message { get; set; }

        [JsonProperty("target")]
        public string Target { get; set; }

        [JsonProperty("details"]
        public ErrorResponse[] Details { get; set; }
    }

If we don't contain "details" then user will get a partiall error response.

Comment (VeryEarly) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: I add ErrorResponseCommon below to have the "details" field, and the definition should be identical with the one defined in common-types.

BTW I don't have code generation environment setup in this laptop, will verify it manually with this new added block tomorrow once I got to the office.

Comment (arolshan) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: ok

**Comment by arolshan** [comment]
Hi wouldn't it be easier to just search & replace to common types ?

**Review by arolshan — APPROVED** [review]

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json
Comment (majastrz) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: >ErrorResponseCommon [](start = 37, length = 19)

ErrorResponseCommon declares a response with code, message, etc. properties which is not correct. Take a look at [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content). The top-level property should be `error`, which is an object that contains `code`, `message`, and other properties.

**Comment by majastrz** [comment]
I added some comments. Please take a look.

**Comment by majastrz** [comment]
I added a comment about the error contract. Please take a look.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json
Comment (VeryEarly) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json:56]: this definition of ErrorResponseCommon was copied from https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/types.json#L272

Can you confirm whether this definition need to be updated?

Comment (majastrz) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json:56]: No, not exactly. The issue is that you are making your error response like this:
```
{
  code: "",
  message: ""
}
```
It needs to be like this:
```
{
  error: {
    code: "",
    message: "",
    ....
  }
}
```

So the definition in common-types is correct if you use it as the schema of the "error" property.

---
In reply to: [401315210](https://github.com/Azure/azure-rest-api-specs/pull/8819#discussion_r401315210) [](ancestors = 401315210)

**Review by majastrz — APPROVED** [review]
Signed off from ARM side. You should look into adding `default` responses to other paths, but that seems like a pre-existing condition, so I won't block on that.

**Comment by weidongxu-microsoft** [comment]
@VeryEarly 

There is lots of rename etc. that causing breaking change. I know you can do this on preview. But would you care to have a quick check on the BreakingChange CI and let us know whether all of these are intended.

I will need to pass it to management for merge, since CI fail on BreakingChange.

**Comment by VeryEarly** [comment]
> @VeryEarly
> 
> There is lots of rename etc. that causing breaking change. I know you can do this on preview. But would you care to have a quick check on the BreakingChange CI and let us know whether all of these are intended.
> 
> I will need to pass it to management for merge, since CI fail on BreakingChange.

Hi @weidongxu-microsoft ,

These rename are for unblock SDK generation fail for dotnet due to inconsistent naming convention.

These breaking change looks good except 
"The new version has a different 'pattern' value than the previous one."

Sorry I'm not familiar with swagger spec rules but seems like I didn't change such property "pattern", can you help me understand what is this line for？

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json
Comment (arolshan) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json]: marcin you are **wrong**. 

Please see documentation for azure async operation response
[https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations#azure-asyncoperation-request-and-response](url)

This is the right contract:
```

{
    "id": "{resource path from GET operation}",
    "name": "{operation-id}", 
    "status" : "Succeeded | Failed | Canceled | {resource provider values}", 
    "startTime": "2017-01-06T20:56:36.002812+00:00",
    "endTime": "2017-01-06T20:56:56.002812+00:00",
    "percentComplete": {double between 0 and 100 },
    "properties": {
        /* Specific resource provider values for successful operations */
    },
    "error" : { 
        "code": "{error code}",  
        "message": "{error description}" 
    }
}

```
what  you're suggesting is **wrong** and against the documentation. You suggested the following:
```

{
    "id": "{resource path from GET operation}",
    "name": "{operation-id}", 
    "status" : "Succeeded | Failed | Canceled | {resource provider values}", 
    "startTime": "2017-01-06T20:56:36.002812+00:00",
    "endTime": "2017-01-06T20:56:56.002812+00:00",
    "percentComplete": {double between 0 and 100 },
    "properties": {
        /* Specific resource provider values for successful operations */
    },
    "error" : { 
        "error": {
            "code": "{error code}",  
            "message": "{error description}" 
       }
    }
}

```

Please also see the master. veryealy doesn't do any breaking change here. He merely changed naming.

**Comment by weidongxu-microsoft** [comment]
@VeryEarly

It appears to be caused by repalcing "../../../../../common-types/resource-management/v1/types.json#/parameters/ResourceGroupNameParameter" with "#/parameters/SubscriptionIdParameter".

The one in common-types got `"pattern": "^[-\\w\\._\\(\\)]+$",` defined.

https://dev.azure.com/azure-sdk/public/_build/results?buildId=344704&view=logs&j=08b5038d-1e84-5877-88da-859df8312b69&t=c1978ecd-fc8a-5291-6d7a-7dd459d8b50f&l=439

**Comment by VeryEarly** [comment]
> @VeryEarly
> 
> It appears to be caused by replacing "../../../../../common-types/resource-management/v1/types.json#/parameters/ResourceGroupNameParameter" with "#/parameters/SubscriptionIdParameter".
> 
> The one in common-types got `"pattern": "^[-\\w\\._\\(\\)]+$",` defined.
> 
> https://dev.azure.com/azure-sdk/public/_build/results?buildId=344704&view=logs&j=08b5038d-1e84-5877-88da-859df8312b69&t=c1978ecd-fc8a-5291-6d7a-7dd459d8b50f&l=439

Got it, so the reason to replace "../../../../../common-types/resource-management/v1/types.json#/parameters/ResourceGroupNameParameter" with "#/parameters/SubscriptionIdParameter" is because other specs need to be included in the same tag, they have "#/parameters/SubscriptionIdParameter" defined and used already which conflict with "../../../../../common-types/resource-management/v1/types.json#/parameters/ResourceGroupNameParameter".

These breaking changes are intended, thanks.

**Comment by weidongxu-microsoft** [comment]
@VeryEarly 

Got it. Please let @akning-ms know when you want to merge to master.

**Pull request merged**
